### PR TITLE
Allow for custom uid attribute in passwd ldap source

### PR DESCRIPTION
**Pull request by kev009**
We store our POSIX usernames in a different field than uid.

This is a simple patch just to gauge how this should be done, I'm open to feedback on how you want it handled.

### CODE REVIEW

**Comment by jaqx0r** [comment]
Looks reasonable :)

I wonder if the config should be passed in by default to the UpdateGetter base class __init__(), so that self.conf is availabe to all subclasses.  What do you think -- would other update getters want to modify behaviour from the config like this?

Can you please document uidattr in the config manpage and put an example usage in the example config as well?

**Comment by kev009** [comment]
Not sure how to do the base class change.

Docs added

I added another two options to regex the uid and group member/memberOf because unfortunately we treat our cn/dn different than our uids.  Does this seem okay?

File: nss_cache/sources/ldapsource.py [hunk 1]
```diff
@@ -566,10 +579,13 @@ class GroupUpdateGetter(UpdateGetter):
 
   def __init__(self,conf):
     super(GroupUpdateGetter, self).__init__()
-    if conf.has_key('rfc2307bis') and conf['rfc2307bis']:
+    self.conf = conf
+    if 'rfc2307bis' in conf and conf['rfc2307bis']:
```
Comment (CrackerJackMack) [inline nss_cache/sources/ldapsource.py]:
```python
if conf.get('rfc2307bis'):
```
Defaults to None, which is false.

File: nss_cache/sources/ldapsource.py [hunk 2]
```diff
@@ -566,10 +579,13 @@ class GroupUpdateGetter(UpdateGetter):
...
       self.attrs = ['cn', 'gidNumber', 'member']
     else:
       self.attrs = ['cn', 'gidNumber', 'memberUid']
+    if 'groupregex' in conf:
```
Comment (CrackerJackMack) [inline nss_cache/sources/ldapsource.py]:
```python
if conf.get('groupregex'):
```

File: nss_cache/sources/ldapsource.py [hunk 3]
```diff
@@ -586,11 +602,17 @@ def Transform(self, obj):
     gr.passwd = '*'
     members = []
     if 'memberUid' in obj:
-      members.extend(obj['memberUid'])
+      if hasattr(self, 'groupregex'):
```
Comment (CrackerJackMack) [inline nss_cache/sources/ldapsource.py]:
```python
if getattr(self, 'groupregex', None):
```
`getattr` doesn't have a default... by default :sunglasses: but can be supplied one.

File: nss_cache/sources/ldapsource.py [hunk 4]
```diff
@@ -586,11 +602,17 @@ def Transform(self, obj):
...
+        members.extend(''.join([x for x in self.groupregex.findall(obj['memberUid'])]))
+      else:
+        members.extend(obj['memberUid'])
     elif 'member' in obj:
       for member_dn in obj['member']:
         member_uid = member_dn.split(',')[0].split('=')[1]
-        members.append(member_uid)
+        if hasattr(self, 'groupregex'):
```
Comment (CrackerJackMack) [inline nss_cache/sources/ldapsource.py]:
```python
if getattr(self, 'groupregex', None):
```

File: nss_cache/sources/ldapsource.py [hunk 5]
```diff
@@ -519,10 +520,15 @@ def GetUpdates(self, source, search_base, search_filter,
 class PasswdUpdateGetter(UpdateGetter):
   """Get passwd updates."""
 
-  def __init__(self):
+  def __init__(self, conf):
     super(PasswdUpdateGetter, self).__init__()
+    self.conf = conf
     self.attrs = ['uid', 'uidNumber', 'gidNumber', 'gecos', 'cn',
                   'homeDirectory', 'loginShell', 'fullName']
+    if 'uidattr' in self.conf:
```
Comment (CrackerJackMack) [inline nss_cache/sources/ldapsource.py]:
```python
if self.conf.get('uidattr'):
```

File: nss_cache/sources/ldapsource.py [hunk 6]
```diff
@@ -519,10 +520,15 @@ def GetUpdates(self, source, search_base, search_filter,
...
+      self.attrs.append(self.conf['uidattr'])
+    if 'uidregex' in self.conf:
```
Comment (CrackerJackMack) [inline nss_cache/sources/ldapsource.py]:
```python
if self.conf.get('uidregex'):
```

**Comment by CrackerJackMack** [comment]
Made some suggestions to clean up the conditionals.  Feels kind of pedantic but also simplifies them to perform two functions in one statement.  Both `X in Y` and 'bool(Y[X] ) == True'.  I missed a few but I think you get the point.

**Comment by jaqx0r** [comment]
Re base class: in __init__() you added conf as a parameter, and save it in
self.conf; I'm suggesting you pass the parameter to super.__init__() and in
UpdateGetter.__init__() you save the value of conf.

File: nss_cache/sources/ldapsource.py [hunk 7]
```diff
@@ -519,10 +520,15 @@ def GetUpdates(self, source, search_base, search_filter,
 class PasswdUpdateGetter(UpdateGetter):
   """Get passwd updates."""
 
-  def __init__(self):
+  def __init__(self, conf):
     super(PasswdUpdateGetter, self).__init__()
```
Comment (jaqx0r) [inline nss_cache/sources/ldapsource.py]: i.e. super(..).__init__(conf)

and in line 423 add a new __init__ method that accepts conf and move line 525 below into this new method.

File: nss_cache/sources/ldapsource.py (hunk 1 above)
Comment (kev009) [inline nss_cache/sources/ldapsource.py]: That's good right?  It's not supposed to be default.  ldap_rfc2307bis = 1 must be set in the config.

Comment (kev009) [inline nss_cache/sources/ldapsource.py]: Sorry, dense today.. use the default None of .get().

**Comment by kev009** [comment]
Ok, I'm happy with this, good now?

**Comment by jaqx0r** [comment]
Brilliant, thanks!

@CrackerJackMack has some good suggestions for cleaning up conf attribute getters, using conf.get instead of the dictionary interface, but I can do them in a followup change.

**Pull request merged**